File: test/conformance/js/rows/E8.mjs
// row js/E8 — "Append and deprecate under the backward-read contract"
// (docs/roadmap.sexp js/E8; docs/SPEC-TABLES.md:116 "by appending and
// deprecating in place").
//
// THE LAW, docs/SPEC-TABLES.md:116: a fixed table is versioned "by appending
// and deprecating in place". The backward-read contract (§5) says a newer
// reader (whose schema has appended fields or deprecated ones) must read an
// older file: appended fields are answered by the prefill (their defaults),
// and a deprecated field is read but its value means nothing.
//
// DERIVATION OF THE BYTE VECTORS FROM THE LAW:
//
// A fixed-form layout is a u32 entry-count followed by 17-byte entries
// (docs/FIXED-FORM-ALGORITHM.md §1.1). The root is kind 13 (table), its
// children are the fields in pre-order.
//
// The "old" writer has two int32 fields {a, b}. The "new" reader has three:
// {a, b, c} with c appended. Both layouts are valid; the reader compiles a
// plan from the old to the new, and the plan copies a and b while c is left
// to the prefill.
//
// FNV-1a 64-bit of "a": id_lo = 0x8601ec8c, id_hi = 0xaf63dc4c
// FNV-1a 64-bit of "b": id_lo = 0x8601f1a5, id_hi = 0xaf63df4c
// FNV-1a 64-bit of "c": id_lo = 0x8601eff2, id_hi = 0xaf63de4c
// (the exact values do not matter for the plan — only equality does).
// Kind 4 = int32, width 4, children 0. Kind 13 = table.
//
// Old layout bytes (3 entries = 55 bytes):
//   header: count=3
//   entry 0: root table, size=8, children=2
//   entry 1: a, kind=4, size=4, children=0
//   entry 2: b, kind=4, size=4, children=0
//
// New layout bytes (4 entries = 72 bytes):
//   header: count=4
//   entry 0: root table, size=12, children=3
//   entry 1: a (same id, kind, size)
//   entry 2: b (same id, kind, size)
//   entry 3: c (new id, kind=4, size=4, children=0)

import { resolve } from "node:path";
import { pathToFileURL } from "node:url";

const generated = process.env.SCHEMA_JS_GENERATED ?? "build/tables-generated-js";
const load = (p) => import(pathToFileURL(resolve(generated, p)).href);

const tables = await load("examples/TabledemoTable.js");

const {
  TableFixedLayoutView,
  TableFixedParseLayout,
  TableFixedCompile,
  TableFixedPlan,
  TableFixedReport,
  TableFixedRun,
  TableFixedHoles,
} = tables;

// The prefill bytes for a 12-byte body of three int32s at defaults (all zero).
const prefill = new Uint8Array(12); // zero defaults

// The reader's dst offsets for the new layout: 5 lanes per entry,
// lane 0 = storage offset (dst).
// Entry 0: root at 0
// Entry 1: field a at offset 0
// Entry 2: field b at offset 4
// Entry 3: field c at offset 8
const newDst = new Int32Array([
  0, 0, 0, 0, 0,
  0, 0, 0, 0, 0,
  4, 0, 0, 0, 0,
  8, 0, 0, 0, 0,
]);

// ---- old layout: root {a:int32, b:int32} ----
const oldLayout = new Uint8Array([
  0x03, 0x00, 0x00, 0x00,                                   // count = 3
  0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,           // root id lo,hi
  0x0d, 0x08, 0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x00,     // kind=13(table), size=8, children=2
  0x10, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,           // field a id
  0x04, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,     // kind=4(int32), size=4, children=0
  0x20, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,           // field b id
  0x04, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,     // kind=4(int32), size=4, children=0
]);

// ---- new layout: root {a:int32, b:int32, c:int32} — c appended ----
const newLayout = new Uint8Array([
  0x04, 0x00, 0x00, 0x00,                                   // count = 4
  0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,           // root id lo,hi
  0x0d, 0x0c, 0x00, 0x00, 0x00, 0x03, 0x00, 0x00, 0x00,     // kind=13(table), size=12, children=3
  0x10, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,           // field a id (same)
  0x04, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,     // kind=4(int32), size=4, children=0
  0x20, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,           // field b id (same)
  0x04, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,     // kind=4(int32), size=4, children=0
  0x30, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,           // field c id (new)
  0x04, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,     // kind=4(int32), size=4, children=0
]);

let failed = false;
function check(ok, what) {
  console.log((ok ? "ok  " : "FAIL") + " " + what);
  if (!ok) { failed = true; }
}

// Parse both layouts.
const oldView = new TableFixedLayoutView();
check(TableFixedParseLayout(oldLayout, 0, oldLayout.length, oldView),
  "old layout parses: root {a,b}, " + oldView.count + " entries");

const newView = new TableFixedLayoutView();
check(TableFixedParseLayout(newLayout, 0, newLayout.length, newView),
  "new layout parses: root {a,b,c}, " + newView.count + " entries");

// Compile a plan from the OLD layout to the NEW layout (the backward-read
// direction: newer reader, older writer).
const plan = new TableFixedPlan(64, 12, 64);
const census = new TableFixedReport();
const entries = TableFixedCompile(oldView, newLayout, newDst, plan, census);
check(entries >= 0, "plan compiles from old→new (" + entries + " entries, split=" + plan.split + ")");

// ---- THE APPEND ASSERTION ----
// The old writer's body: a=42, b=99 (8 bytes, little-endian).
const oldBody = new Uint8Array([
  0x2a, 0x00, 0x00, 0x00,  // a = 42
  0x63, 0x00, 0x00, 0x00,  // b = 99
]);

// The plan's image is the reader's storage. Prefill first (what the plan
// does not write), then run.
plan.image.fill(0); // reader's defaults: zero
const holes = TableFixedHoles(plan.entries, entries, plan.cover, plan.holes);
for (let h = 0; h < holes; h++) {
  const o = plan.holes[h * 2], z = plan.holes[h * 2 + 1];
  for (let i = 0; i < z; i++) { plan.image[o + i] = prefill[o + i]; }
}

const srcView = new DataView(oldBody.buffer);
plan.split = entries;
plan.count = entries;

const report = new TableFixedReport();
TableFixedRun(plan.entries, entries, oldBody, srcView, 0, plan.image, plan.view, plan.remap, report);

const imageView = plan.view;
check(imageView.getInt32(0, true) === 42,
  "E8 append: a lands as 42 (got " + imageView.getInt32(0, true) + ")");
check(imageView.getInt32(4, true) === 99,
  "E8 append: b lands as 99 (got " + imageView.getInt32(4, true) + ")");
check(imageView.getInt32(8, true) === 0,
  "E8 append: appended c lands as 0 default (got " + imageView.getInt32(8, true) + ")");

// No counters fire: the old writer named everything the new reader knows.
check(report.unknown === 0 && report.clamped === 0 && report.kindMismatch === 0,
  "E8 append: no counter fires (unknown=" + report.unknown +
  ", clamped=" + report.clamped + ", kindMismatch=" + report.kindMismatch + ")");
check(!report.malformed && report.refused === 0,
  "E8 append: no refusal or malformed (refused=" + report.refused +
  ", malformed=" + report.malformed + ")");

// ---- THE DEPRECATION HALF: same layout, different meaning ----
// A deprecated field does NOT change the layout — the slot stays where it is.
// The old writer writes it; the new reader reads it. The app ignores it.
// We prove this by running the plan in the other direction: the "new" writer
// (with c) is read by the "old" reader (without c).
const newBody = new Uint8Array([
  0x0a, 0x00, 0x00, 0x00,  // a = 10
  0x14, 0x00, 0x00, 0x00,  // b = 20
  0x1e, 0x00, 0x00, 0x00,  // c = 30 (appended, unknown to old reader)
]);

// Old reader's image: 8 bytes
const oldImage = new Uint8Array(8);
oldImage.fill(0);
const oldImageView = new DataView(oldImage.buffer);

// Plan from new to old: old reader does not know c.
const oldDst = new Int32Array([
  0, 0, 0, 0, 0,
  0, 0, 0, 0, 0,
  4, 0, 0, 0, 0,
]);

const plan2 = new TableFixedPlan(64, 8, 64);
const census2 = new TableFixedReport();
const entries2 = TableFixedCompile(newView, oldLayout, oldDst, plan2, census2);
check(entries2 >= 0, "plan compiles from new→old (" + entries2 + " entries)");

plan2.image.fill(0);
const holes2 = TableFixedHoles(plan2.entries, entries2, plan2.cover, plan2.holes);
for (let h = 0; h < holes2; h++) {
  const o = plan2.holes[h * 2], z = plan2.holes[h * 2 + 1];
  for (let i = 0; i < z; i++) { plan2.image[o + i] = 0; }
}

plan2.split = entries2;
plan2.count = entries2;

const newSrcView = new DataView(newBody.buffer);
const report2 = new TableFixedReport();
TableFixedRun(plan2.entries, entries2, newBody, newSrcView, 0, plan2.image, oldImageView, plan2.remap, report2);

check(oldImageView.getInt32(0, true) === 10,
  "E8 deprecation: old reader gets a=10 (got " + oldImageView.getInt32(0, true) + ")");
check(oldImageView.getInt32(4, true) === 20,
  "E8 deprecation: old reader gets b=20 (got " + oldImageView.getInt32(4, true) + ")");

// c is unknown to the old reader — one unknown per peer (§5.4), counted at
// compile time and carried on the plan (the census, §5.9 #6).
check(census2.unknown === 1,
  "E8 deprecation: one unknown for appended c (census got " + census2.unknown + ")");
check(!report2.malformed && report2.refused === 0,
  "E8 deprecation: no refusal or malformed (refused=" + report2.refused +
  ", malformed=" + report2.malformed + ")");

if (failed) {
  console.log("FAILED");
  process.exit(1);
}
console.log("E8: append and deprecate under the backward-read contract — the plan copies known fields and defaults the rest, a deprecated slot is read but its value means nothing");

File: test/conformance/dart/rows/E8.dart
// CARD: cell-dart-field-evolution E8 — append and deprecate under the
// backward-read contract
// LAW: docs/SPEC-TABLES.md:116 — a fixed table is "versioned by appending and
// deprecating in place."; docs/SPEC-TABLES.md:16416 — "A newer reader reads
// every older file. An older reader given a newer file refuses it by name,
// before any record, and reads nothing."; docs/SPEC-TABLES.md:3171 — "A fixed
// table evolves APPEND-ONLY. A new field goes at the BOTTOM and nowhere else.
// A field that has outlived its use is DEPRECATED IN PLACE — it keeps its
// slot forever."
// CELL: dart/E8
// TEST: In the generated fixed-form tables code the appended field rides the
// new generation's versioned layout under its own declared identity and the
// old generation has no such slot; and the backward-read contract holds — an
// older reader refuses the newer file (the file that carries the appended
// slot) BY NAME, before any record, reading nothing, counters unmoved, the
// caller's poisoned storage untouched, the file's own hash reported.
//
// V1/V2 (test/tables/V1.schema, V2.schema) are the evolution pair: V2 carries
// `c bool = true` where V1 has no such field.
//
// DERIVATIONS (every constant this test pins):
// - fnv1a64: the wire id is the hash of the wire name (docs/SPEC-TABLES.md §5)
//   — offset basis 0xCBF29CE484222325, prime 0x100000001B3, one byte at a
//   time, as spelled in ir/tablewire.go TableWireId.
// - A LAYOUT is a u32 entry count then 17-byte entries: u64 id, u8 kind, u32
//   size, u32 children, all little-endian (TableFixedLimits.entryBytes = 17 in
//   the generated runtime; the walk is a pre-order in the writer's declared
//   order).
// - kind 1 is a bool (the frozen §3 kind vocabulary, docs/SPEC-TABLES.md §3);
//   a bool's slot is one byte.
// - refusal 15 is `layout_newer`: the integers are each leg's own, taken in
//   the order §5.3 names the refusals, and this leg's 15 agrees with the JS
//   leg's TableFixedRefusal.LayoutNewer = 15 (internal/codegen/jstable/
//   fixedruntime.go:135); the NAME is the contract.
// - 0x0EADBEEFCAFEF00D is a forged header hash no lineage entry holds, the
//   versioning gate's hash_unknown case.
// - the 0x5A POISON is §5.7/§5.9 #17's discipline: a refusal that ran a decode
//   step would leave a mark in the caller's storage.

import 'dart:typed_data';
import 'dart:io';

import '../../../../build/tables-generated-dart/v1/V1Fixed.dart' as v1;
import '../../../../build/tables-generated-dart/v1/Tblv1Fixed.dart' as v1rt;
import '../../../../build/tables-generated-dart/v2/V2Fixed.dart' as v2;
import '../../../../build/tables-generated-dart/v2/Tblv2Fixed.dart' as v2rt;

/// fnv1a64 of a wire name, as ir.TableWireId computes it.
int fnv1a64(String name) {
  var h = 0xCBF29CE484222325;
  for (var i = 0; i < name.length; i++) {
    h ^= name.codeUnitAt(i) & 0xFF;
    h *= 0x100000001B3;
  }
  return h;
}

/// Every layout entry whose id is `want`, decoded off the 17-byte rows.
List<int> entriesWithId(Uint8List layout, int want) {
  final view = ByteData.sublistView(layout);
  final count = view.getUint32(0, Endian.little);
  final found = <int>[];
  for (var i = 0; i < count; i++) {
    final at = 4 + i * 17;
    if (view.getUint64(at, Endian.little) == want) {
      found.add(i);
    }
  }
  return found;
}

bool entryIsBoolSlot(Uint8List layout, int entry) {
  final view = ByteData.sublistView(layout);
  final at = 4 + entry * 17;
  return layout[at + 8] == 1 && // TableKindBool, the frozen §3 code
      view.getUint32(at + 9, Endian.little) == 1 && // one byte
      view.getUint32(at + 13, Endian.little) == 0;
}

void main() {
  var failed = false;

  void check(bool ok, String msg) {
    print('${ok ? 'ok' : 'FAIL'}: $msg');
    if (!ok) {
      failed = true;
    }
  }

  // ---- THE APPENDED FIELD RIDES THE NEW GENERATION'S LAYOUT ----
  final cId = fnv1a64('c');
  final v2C = entriesWithId(v2.cfgFixedLayout, cId);
  check(
    v2C.length == 1 && entryIsBoolSlot(v2.cfgFixedLayout, v2C.first),
    'E8: v2\'s Cfg layout carries exactly ONE entry for the appended field '
    '`c` — kind=1 (bool), one byte',
  );
  check(
    entriesWithId(v1.cfgFixedLayout, cId).isEmpty,
    'E8: v1\'s Cfg layout carries no such slot — the old writer could not '
    'have written one',
  );

  // ---- THE BACKWARD-READ CONTRACT ----
  // One V2 record, carrying the appended slot at a NON-default value, offered
  // to the OLD build's reader.
  final one = v2rt.Cfg();
  one.title.setRange(0, 5, 'hello'.codeUnits);
  one.titleLength = 5;
  one.c = false; // c's declared default is true
  final file = Uint8List(v2.cfgFixedMeasure(1));
  check(
    v2.cfgFixedSave(<v2rt.Cfg>[one], 1, file) == file.length,
    'E8: v2 saves its record — a file whose layout carries the appended slot',
  );

  // THE POISON: the caller's record storage, filled before the load.
  final plan = v1.cfgFixedNewPlan();
  plan.image.fillRange(0, plan.image.length, 0x5A);

  final back = <v1rt.Cfg>[v1rt.Cfg()];
  final r = v1rt.TableFixedReport();
  final n = v1.cfgFixedLoad(back, 1, file, file.length, plan, r);

  // REFUSED BY NAME, before any record: `layout_newer`, the older build's
  // answer to a newer file. The integer 15 is this leg's own spelling of the
  // §5.3 name (the JS leg agrees); the name is the contract.
  check(n == -1, 'E8: the older reader refuses the newer file (n=$n)');
  check(r.refused == 15, 'E8: the refusal is 15, the legs\' shared spelling '
      'of layout_newer (got ${r.refused})');
  check(
    v1rt.TableFixedRefusal.name(r.refused) == 'layout_newer',
    'E8: refused BY NAME — ${v1rt.TableFixedRefusal.name(r.refused)}',
  );
  check(!r.malformed, 'E8: a refusal by name is not damage (malformed=false)');

  // REFUSE IS TOTAL: no counter moved, and the poisoned storage is untouched —
  // the refusal happened before any record, so nothing was decoded.
  check(
    r.unknown == 0 &&
        r.kindMismatch == 0 &&
        r.clamped == 0 &&
        r.widened == 0 &&
        r.duplicate == 0,
    'E8: refuse moves no counter (unknown=${r.unknown} kind=${r.kindMismatch} '
    'clamped=${r.clamped} widened=${r.widened} duplicate=${r.duplicate})',
  );
  var poison = true;
  for (var i = 0; i < plan.image.length; i++) {
    if (plan.image[i] != 0x5A) {
      poison = false;
      break;
    }
  }
  check(
    poison,
    'E8: the reader READS NOTHING — every poisoned byte of the caller\'s '
    'record storage is still 0x5A',
  );

  // THE REFUSAL CARRIES THE FILE'S HASH — read back from the bytes themselves.
  final fileHash = ByteData.sublistView(
    file,
  ).getUint64(8, Endian.little);
  check(
    r.layoutHash == fileHash,
    'E8: the refusal reports the FILE\'S hash '
    '(0x${r.layoutHash.toRadixString(16)})',
  );

  // A STRANGER'S HASH: a forged header hash no lineage entry holds is refused
  // the same way, and reports the FORGED value.
  ByteData.sublistView(
    file,
  ).setUint64(8, 0x0EADBEEFCAFEF00D, Endian.little);
  final plan2 = v1.cfgFixedNewPlan()..image.fillRange(0, plan.image.length, 0x5A);
  final r2 = v1rt.TableFixedReport();
  final n2 = v1.cfgFixedLoad(back, 1, file, file.length, plan2, r2);
  check(
    n2 == -1 &&
        r2.refused == 15 &&
        v1rt.TableFixedRefusal.name(r2.refused) == 'layout_newer' &&
        r2.layoutHash == 0x0EADBEEFCAFEF00D,
    'E8: a hash in no lineage entry refuses layout_newer and reports the '
    'file\'s (forged) hash',
  );
  var poison2 = true;
  for (var i = 0; i < plan2.image.length; i++) {
    if (plan2.image[i] != 0x5A) {
      poison2 = false;
      break;
    }
  }
  check(poison2, 'E8: the stranger-hash refusal reads nothing either');

  if (failed) {
    exit(1);
  }
  print('E8: append and deprecate under the backward-read contract — PASS');
}
